EIPS/eip-1965.md

```diff
@@ -0,0 +1,60 @@
+---
+eip: 1965
+title: Valid ChainID with blockNumber
+author: Ronan Sandford (@wighawag)
+category: Core
+type: Standards Track
+discussions-to: https://ethereum-magicians.org/t/eip-1965-valid-chainid-for-specific-blocknumber-protect-all-forks/3181
+status: Draft
+created: 2019-04-20
+---
+
+## Abstract
+This EIP adds an precompile that returns whether a specific chainID (EIP-155 unique identifier) is valid at a specific blockNumber. ChainID are assumed to be valid up to the blockNumber at which they get replaced by a new chainID.
+
+## Motivation
+[EIP-155](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md) proposes to use the chain ID to prevent replay attacks between different chains. It would be a great benefit to have the same possibility inside smart contracts when handling signatures, especially for Layer 2 signature schemes using [EIP-712](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md).
+
+[EIP-1344](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1344.md) is attempting to solve this by giving smart contract access to the tip of the chainID history. This is insuficient as such value is changing. Hence why EIP-1344 describes a contract based solution to work arround the problem. It would be better to solve it in a simpler, cheaper and safer manner, removing the potential risk of misuse present in EIP-1344. Furthermore EIP-1344 can't protect replay properly for minority-led hardfork as the caching system cannot guarantee accuracy of the blockNumber at which the new chainID has been introduced.
+
+[EIP-1959](https://github.com/ethereum/EIPs/pull/1959) solves the issue of EIP-1344 but do not attempt to protect from minority-led hardfork as mentioned in the rationale. We consider this a mistake, since it remove some freedom to fork. We consider that all fork should be given equal oportunities. And while there will always be issues we can't solve for the majority that ignore a particular fork, **users that decide to use both the minority-fork and the majority-chain should be protected from replay without having to wait for the majority chain to update its chainID.**
+
+## Specification
+Adds a new precompile which uses 2 argument : a uint256 value that represent the chainID to test and a uint256 value representing the blockNumber at which the chainID is tested. It return ```true```  if the chainID is valid at the specific blockNumber, ```false``` otherwise. Note that chainID are considered valid up to the blockNumber at which they are replaced. So they are valid for every blockNumber past their replacement.
+
+The operation costs `G_blockhash` + `G_verylow` to execute.
```

Comment:
Can some client developer comment how feasible is this? Where is the data stored?

cc @karalabe @holiman @folsen @chfast @gumb0

Reply:
Up until now, the chain id has never changed. It's in a config that's basically 'everywhere', so should be fairly simple to serve to the EVM. 
However, what's the point in providing a chainid to test and a blocknumber, instead of just a blocknumber and get back the chain id?

Reply:
One reason is that, as mentioned in the proposal, chainIds are considered valid up to the blockNumber at which they are replaced. So at a specific blockNumber multiple chainIds are valid. 

In particular as it stands
```validChainId(<chainId>, 0) ``` return true of all chainId part of the history.

This is so that the method is more flexible for application that do not want to care about the block number, since discussion is still happening whether we should allow for such applications.

In particular, unless EIP-712 adapt to enforce blockNumber (which it should in my opinion) applications are unlikely to use blockNumber.

In other words, it make the proposal work in both situation. 

I think the community need to take a stance on whether we should support minority-led hardfork (I personally think we should since without it we loose a big part of the freedom to fork, hence decentralization). If that is the case, we need to adapt #712  as mentioned in the rationale and we could then make it so that a chainId is invalid when coupled with blockNumber past its introduction, forcing application smart contract to consider the blockNumber for the benefit of all minority-led forks.

The current API does not need to change though.

Another reason is simply that there is no need for accessing the actual chainId except for comparison, which this method do for you

Reply:
Let's continue the discussion on the discussion thread : https://ethereum-magicians.org/t/eip-1965-valid-chainid-for-specific-blocknumber-protect-all-forks/3181